### PR TITLE
Handle human readable number casting in general

### LINKED ISSUE #1309: Handle human readable number casting in general
## Problem

Currently, casting from textual types (e.g., `TEXT`) to numbers relies on the default PostgreSQL number handling. This will be solved for the `NUMERIC` type once #1107 is solved, but we should also handle other number types.

## Proposed solution

We should write custom logic to allow handling casting more strings to number types.

- Casting to number types with decimals should use the same (or very similar) logic to casting to `NUMERIC`.
- Casting to number types without decimals (e.g., `INTEGER`) should use a subset of that logic that handles only grouping separators.

## Additional context

This was originally inspired by the custom casting work for money types. See #1137 . This issue should complete the work begun when solving #1107 .

### PR DESCRIPTION
**Pull request by Aditramesh**
Fixes #1309

Adds functionalities to handle casting of text to integer and decimal types which was implemented for numeric types in #1355

**Technical details**
Uses similar logic for the integer types and added functions for decimal types(real ,double precision ) to work with the logic for numeric 

**Screenshots**

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [ ] My pull request targets the `develop` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by dmos62** [comment]
Continuation of https://github.com/centerofci/mathesar/pull/2465

**Review by mathemancer — CHANGES_REQUESTED** [review]
There's an issue with the integer-finding regex that I found.

Also, please remember to write some tests. Parsing regexes by eyeball is not a robust way for me to check that these have the correct behavior, and we won't be merging this PR without a decent test suite associated with it.

Thank you for your effort so far.

File: db/types/operations/cast.py [hunk 1]
```diff
@@ -1027,6 +1062,95 @@ def _build_numeric_array_function():
...
+
+def _build_integer_array_function():
+    """
+    The main reason for this function to be separate is for testing. This
+    does have some performance impact; we should consider inlining later.
+    """
+    qualified_function_name = get_qualified_name(INTEGER_ARR_FUNC_NAME)
+
+    single_digit = r"^[0-9]$"
+    no_separator = r"[0-9]{2,}(?:([,])[0-9]{1,2}|[0-9]{4,})?"
```
[the remark this answers is not in the document]
Comment (Aditramesh) [reply db/types/operations/cast.py:1105]: I thought numbers like 12,345 also have to be matched,that's why i was writing those two regex's ,later i realised 12,345 should not be matched so i think these the regex's should look like this now 
    no_separator = r"^-?\d+$"
    comma_separator = r"[0-9]{1,3}(?:(,)[0-9]{3}){2,}"
    period_separator = r"[0-9]{1,3}(?:(\.)[0-9]{3}){2,}"
    comma_separator_lakh_system = r"[0-9]{1,2}(?:(,)[0-9]{2})+,[0-9]{3}?"
    single_quote_separator = r"[0-9]{1,3}(?:(\'')[0-9]{3})+"
    space_separator = r"[0-9]{1,3}(?:( )[0-9]{3})+(?:([,])[0-9]+)?"

[the remark this answers is not in the document]
Comment (mathemancer) [reply db/types/operations/cast.py:1105]: You don't need the beginning or end string markers (or the `[-]` in the individual expression. Those are handled by the wrapping expression.

[the remark this answers is not in the document]
Comment (Aditramesh) [reply db/types/operations/cast.py]: I have made those changes and added tests,thanks for the help throughout @mathemancer .

File: db/types/operations/cast.py [hunk 2]
```diff
@@ -1028,6 +1063,91 @@ def _build_numeric_array_function():
...
+    The main reason for this function to be separate is for testing. This
+    does have some performance impact; we should consider inlining later.
+    """
+    qualified_function_name = get_qualified_name(INTEGER_ARR_FUNC_NAME)
+
+    no_separator = r"\d+"
+    comma_or_period_separator = r"([0-9]|[0-9]{3})(([,.])[0-9]{3})+"
+    comma_separator_lakh_system = r"([0-9]{1,2}(?:(,)[0-9]{2})+),[0-9]{3}"
+    space_separator = r"[0-9]{1,3}(?:( )[0-9]{3})+"
+    single_quote_separator = r"([0-9]{1,3})(?:(\'')[0-9]{3})+(?:(\'')?[0-9]*)?"
```
Comment (mathemancer) [inline db/types/operations/cast.py:1109]: For the comma and period separator ones, we need at least 2 groups of three to be sure that it's an integer. This is crucial in our inference system at the moment. I.e., it's not possible to determine the value of `1,234` without knowing the locale.

**Review by mathemancer — CHANGES_REQUESTED** [review]
Thank you for writing some test cases. It really helps me see what's going on at a glance. However, there's been a regression in the logic since my last review. Please see the specific comment.

Thanks for your effort on this so far!

File: db/types/operations/cast.py (hunk 2 above)
Comment (Aditramesh) [reply db/types/operations/cast.py:1109]: oh sorry my bad ,changing the regex to 
comma_or_period_separator = r"([0-9]|[0-9]{3})(([,.])[0-9]{3}){2,}" should fix the issue you are referring to,correct ?

**Comment by rajatvijay** [comment]
Closing this due to inactivity.

**Comment by Aditramesh** [comment]
@rajatvijay the requested changed were made the pr has to be checked once and merged

**Comment by mathemancer** [comment]
I'm sorry, @Aditramesh . This still isn't working correctly, and I've run out of time for going back-and-forth on this PR. I'm closing it.

I should have done this awhile ago.

**Pull request closed without merging**